src/channel-qvm.lisp

```diff
@@ -0,0 +1,374 @@
+  (loop :for rule in rules
+        :when (matches-rule rule instr-qubits instr-name position)
+          :return (operation-elements rule)))
+
+
+(defmethod apply-all-kraus-operators ((qvm channel-qvm) (instr quil:gate-application) kraus-operators)
+  "Apply all the kraus operators to the state of the system."
+  (loop :for kops :in kraus-operators
+        :do
+           (check-kraus-ops kops)
+           (apply-kraus-ops qvm instr kops)))
+
+
+(defmethod apply-kraus-ops ((qvm channel-qvm) (instr quil:gate-application) kraus-ops)
+  (let* ((amps (%trial-amplitudes qvm))
+         (r (random 1.0d0))
+         (summed-p 0.0d0)
+         (logical-qubits (quil:application-arguments instr))
+         (qubits (mapcar #'quil:qubit-index logical-qubits))) ;; this needs to be some kind of reference to operational qubits. Have another dict? 
+    (check-type amps quantum-state)
+    (loop :for kj :in kraus-ops
+          :do
+             (replace amps (amplitudes qvm))
+             (apply-matrix-operator (magicl-matrix-to-quantum-operator kj) amps (apply #'nat-tuple qubits))
+             (incf summed-p (psum #'probability amps))
+          :until (>= summed-p r))
+
+    (rotatef (amplitudes qvm) (%trial-amplitudes qvm))
+    (normalize-wavefunction (amplitudes qvm))))
+
+
+(defgeneric apply-classical-readout-noisee (qvm instr)
+  (:documentation "Given a QVM and a (measurement) instruction INSTR, corrupt the readout bit according to the POVM specifications of QVM.")
+  ;; Pure state QVM has no readout noise.
+
+  (:method ((qvm pure-state-qvm) (instr quil:measurement))
+    (declare (ignore qvm instr))
+    nil)
+  ;; Readout noise only happens to the resulting classical bit (i.e.,
+  ;; it's classical noise). As such, discarding that bit doesn't
+  ;; warrant any sort of special treatment.
+  (:method ((qvm approx-qvm) (instr quil:measure))
+    (%corrupt-readout qvm instr (readout-povms qvm))
+    )
+  (:method ((qvm channel-qvm) (instr quil:measure-discard))
+    (declare (ignore qvm instr))
+    nil)
+  ;; We do have a readout bit, and we want to corrupt it.
+  (:method ((qvm channel-qvm) (instr quil:measure))
+    (%corrupt-readout qvm instr (readout-povms (noise-model qvm))))
+  ;; For compiled measurements, refer to the source of that
+  ;; instruction.
+  (:method ((qvm channel-qvm) (instr compiled-measurement))
+    (apply-classical-readout-noise qvm (source-instruction instr))))
+
+
+(defun %corrupt-readout (qvm instr povm-map)
+  (check-type instr quil:measure)
+  (let* ((q (quil:qubit-index (quil:measurement-qubit instr)))
+         (a (quil:measure-address instr))
+         (c (dereference-mref qvm a))
+         (povm (gethash q povm-map))
+         )
+    (when povm
+      (destructuring-bind (p00 p01 p10 p11) povm
+        (setf (dereference-mref qvm a)
+              (perturb-measurement c p00 p01 p10 p11))))))
+
+
+;; TODO declare inlinabldee, turn on optimization
+(defun perturb-measurement (actual-outcome p00 p01 p10 p11)
```

Comment:
Duplicate of a function in `noisy-qvm.lisp`. Ditto `check-povm` below. Don't forget to delete these before merging.